### PR TITLE
implement dot_product, cosine_similarity/matrix for tensor ops; add shape mismatch e…

### LINKED ISSUE #136: Add cosine similarity metric
Follow the implementation below

image

*https://www.linkedin.com/posts/ashvardanian_100x-10000x-numerical-error-reduction-activity-7242002088803663872-kVsx?utm_source=share&utm_medium=member_desktop

### PR DESCRIPTION
**Pull request by Force1ess**
feat: implement vector operations and error handling

Resolves #136

- Add dot_product function for tensor multiplication
- Implement cosine_similarity and cosine_distance metrics
- Add ShapeMismatch error type for tensor dimension validation
- Include comprehensive tests with edge cases for all operations

This supersedes previous work in #138 and #157

### CODE REVIEW

File: crates/kornia-tensor-ops/src/ops.rs [hunk 1]
```diff
@@ -76,6 +76,170 @@ where
...
+/// use kornia_tensor::{Tensor, CpuAllocator};
+/// use kornia_tensor_ops::ops::dot_product;
+///
+/// let a = Tensor::<i32, 1, CpuAllocator>::from_shape_slice([3], &[1, 2, 3], CpuAllocator).unwrap();
+/// let b = Tensor::<i32, 1, CpuAllocator>::from_shape_slice([3], &[4, 5, 6], CpuAllocator).unwrap();
+/// let result = dot_product(&a, &b).unwrap();
+/// assert_eq!(result, 32); // 1*4 + 2*5 + 3*6 = 4 + 10 + 18 = 32
+/// ```
+pub fn dot_product<T, const N: usize, A>(
+    a: &Tensor<T, N, A>,
```
Comment (edgarriba) [inline crates/kornia-tensor-ops/src/ops.rs:106]:
```suggestion
    a: &Tensor<T, 1, A>,
```
The current implementation will work only in case of Tensor1. It’s totally fine but I would specialise the function generics and limit for tensor of rank 1 and call it `dot_product1´. To supports the rest of ranks, I believe there’s the need to add some sort of logic to apply the ops in a specified axis.

File: crates/kornia-tensor-ops/src/ops.rs [hunk 2]
```diff
@@ -76,6 +76,170 @@ where
...
+/// ```
+/// use kornia_tensor::{Tensor, CpuAllocator};
+/// use kornia_tensor_ops::ops::cosine_similarity;
+///
+/// let a = Tensor::<f32, 1, CpuAllocator>::from_shape_slice([3], &[1.0, 0.0, 0.0], CpuAllocator).unwrap();
+/// let b = Tensor::<f32, 1, CpuAllocator>::from_shape_slice([3], &[0.0, 1.0, 0.0], CpuAllocator).unwrap();
+/// let result = cosine_similarity(&a, &b).unwrap();
+/// assert_eq!(result, 0.0); // Orthogonal vectors have cosine similarity of 0
+/// ```
+pub fn cosine_similarity<T, const N: usize, A>(
```
Comment (edgarriba) [inline crates/kornia-tensor-ops/src/ops.rs:158]: Can you check this post ? https://ashvardanian.com/posts/python-c-assembly-comparison/
there some ideas for optimisations here. Besides, can you write a small benchmark scripts (check in other crates the benches). Would be great to see performance number and make it close the ones in the post

**Comment by Force1ess** [comment]
Hi @edgarriba ,
I've implemented the optimized cosine_similarity function as described in the blog post and successfully reproduced comparable performance results. The benchmarks confirm significant speed improvements on my Mac M1 Pro, as shown in the attached screenshots.
One question: Since this optimized implementation doesn't use the dot_product function, should we remove that function from the codebase to reduce redundancy?

image
image

File: crates/kornia-tensor-ops/src/ops.rs [hunk 3]
```diff
@@ -76,6 +76,270 @@ where
...
+/// * `b` - Second tensor
+///
+/// # Returns
+///
+/// A scalar value representing the cosine similarity between the two tensors.
+///
+/// # Errors
+///
+/// If the shapes of the tensors don't match, an error is returned.
+pub fn cosine_similarity_optimized<T, const N: usize, A>(
```
Comment (edgarriba) [inline crates/kornia-tensor-ops/src/ops.rs:263]: Great ! So what do you think about providing a low level kernel function like: `fn cosine_similariy_kernel_float<T>(a: &[T], b: &[T]) -> T`. This will make easy to have a very clean low level api to expose to raw python types (numpy comes with an overhead to pass data from/to) and avoid for now to expose the whole Tensor API. We can do this in a future PR

**Comment by edgarriba** [comment]
> Hi @edgarriba , I've implemented the optimized cosine_similarity function as described in the blog post and successfully reproduced comparable performance results. The benchmarks confirm significant speed improvements on my Mac M1 Pro, as shown in the attached screenshots. One question: Since this optimized implementation doesn't use the dot_product function, should we remove that function from the codebase to reduce redundancy?
> 
> image image

Results are amazing ! I’m pretty we can even optimise a bit more later. I would like to keep the two functions, specially if we can make them very performant as they are very useful for ml in general

File: crates/kornia-tensor-ops/src/ops.rs (hunk 1 above)
Comment (Force1ess) [reply crates/kornia-tensor-ops/src/ops.rs:106]: Hi，given that optimized cosine similarity doesn’t require it, should we consider removing it?
Or we keep it, and I'll fix this.

**Comment by Force1ess** [comment]
> > Hi @edgarriba , I've implemented the optimized cosine_similarity function as described in the blog post and successfully reproduced comparable performance results. The benchmarks confirm significant speed improvements on my Mac M1 Pro, as shown in the attached screenshots. One question: Since this optimized implementation doesn't use the dot_product function, should we remove that function from the codebase to reduce redundancy?
> > image image
> 
> Results are amazing ! I’m pretty we can even optimise a bit more later. I would like to keep the two functions, specially if we can make them very performant as they are very useful for ml in general

Hi, I've finished the improvements we mentioned before.
Here's the newest benchmark result:
image

**Comment by Force1ess** [comment]
By the way, I am pretty interested in optimizing these functions under your guides.
If you have any ideas, please feel free to let me know.

File: crates/kornia-tensor-ops/src/ops.rs (hunk 3 above)
Comment (Force1ess) [reply crates/kornia-tensor-ops/src/ops.rs]: By the way, should I put it in the 'tensor-ops' crate too?

File: crates/kornia-tensor-ops/Cargo.toml [hunk 4]
```diff
@@ -12,6 +12,12 @@ rust-version.workspace = true
 version.workspace = true
 
 [dependencies]
+criterion.workspace = true
```
Comment (edgarriba) [inline crates/kornia-tensor-ops/Cargo.toml:15]: this should go under [dev-dependencies]

File: crates/kornia-tensor-ops/Cargo.toml [hunk 5]
```diff
@@ -12,6 +12,12 @@ rust-version.workspace = true
...
 kornia-tensor = { workspace = true }
 num-traits = { workspace = true }
+rand.workspace = true
```
Comment (edgarriba) [inline crates/kornia-tensor-ops/Cargo.toml:18]: same

File: crates/kornia-tensor-ops/src/ops.rs (hunk 3 above)
Comment (edgarriba) [reply crates/kornia-tensor-ops/src/ops.rs]: yep, create a new `mod kernels` and place it there. This is my suggestion

```rust
fn dot_product1_float_kernel<T>(a: &[T], b: &[T]) -> T
where
    T: Zero + Copy + Clone + std::ops::Add<Output = T> + std::ops::Mul<Output = T>,
{
    let mut result = T::zero();
    for (a_val, b_val) in a.iter().zip(b.iter()) {
        result = result + *a_val * *b_val;
    }
    result
}
```
or
```rust
fn product1_float_kernel<T>(a: &[T], b: &[T]) -> T
where
    T: Zero + Copy + Clone + std::ops::Add<Output = T> + std::ops::Mul<Output = T>,
{
    let result = a
        .iter()
        .zip(b.iter())
        .fold(T::zero(), |acc, (a_val, b_val)| acc + *a_val * *b_val);
    result
}
```
probably the second should leverage better rust internals to vectorise when it's possible

File: crates/kornia-tensor-ops/src/ops.rs [hunk 6]
```diff
@@ -76,6 +76,155 @@ where
...
+
+    // Handle edge cases
+    if magnitude_a == T::zero() || magnitude_b == T::zero() {
+        return Ok(T::zero());
+    }
+
+    // Calculate cosine similarity: dot_product/(sqrt(magnitude_a)*sqrt(magnitude_b))
+    let denominator = magnitude_a.sqrt() * magnitude_b.sqrt();
+    Ok(dot_product / denominator)
+}
```
Comment (edgarriba) [inline crates/kornia-tensor-ops/src/ops.rs:189]: same kernel strategy here 

```rust
fn cosine_similarity_kernel<T>(a: &[T], b: &[T]) -> T
where
    T: num_traits::Float,
{
    let (dot_product, magnitude_a, magnitude_b) = a.iter().zip(b.iter()).fold(
        (T::zero(), T::zero(), T::zero()),
        |(dot_product, magnitude_a, magnitude_b), (a_val, b_val)| {
            let a = *a_val;
            let b = *b_val;
            (
                dot_product + a * b,
                magnitude_a + a * a,
                magnitude_b + b * b,
            )
        },
    );

    if magnitude_a == T::zero() || magnitude_b == T::zero() {
        return T::zero();
    }

    let denominator = magnitude_a.sqrt() * magnitude_b.sqrt();
    dot_product / denominator
}
```

**Comment by edgarriba** [comment]
> By the way, I am pretty interested in optimizing these functions under your guides. If you have any ideas, please feel free to let me know.

lot's of things to do. Curating all this low level kernels needed for ml could be an interesting direction and then efficiently exposing to python

**Comment by Force1ess** [comment]
Hi @edgarriba ,
I've relocated the dependency as per your guides. 
Is there anything else we should address in this PR?

**Comment by Force1ess** [comment]
> > By the way, I am pretty interested in optimizing these functions under your guides. If you have any ideas, please feel free to let me know.
> 
> lot's of things to do. Curating all this low level kernels needed for ml could be an interesting direction and then efficiently exposing to python

While I plan to work on implementing low-level kernels for machine learning.
Maybe I should focus on resolving GSoC-related issues first.

**Comment by edgarriba** [comment]
> > > By the way, I am pretty interested in optimizing these functions under your guides. If you have any ideas, please feel free to let me know.
> > 
> > 
> > lot's of things to do. Curating all this low level kernels needed for ml could be an interesting direction and then efficiently exposing to python
> 
> While I plan to work on implementing low-level kernels for machine learning. Maybe I should focus on resolving GSoC-related issues first.

definitely ! but feel free to open any ticket if you think of some other interesting things to be done later. Maybe other can help too. BTW, are you in the Discord?

**Comment by Force1ess** [comment]
> > > > By the way, I am pretty interested in optimizing these functions under your guides. If you have any ideas, please feel free to let me know.
> > > 
> > > 
> > > lot's of things to do. Curating all this low level kernels needed for ml could be an interesting direction and then efficiently exposing to python
> > 
> > 
> > While I plan to work on implementing low-level kernels for machine learning. Maybe I should focus on resolving GSoC-related issues first.
> 
> definitely ! but feel free to open any ticket if you think of some other interesting things to be done later. Maybe other can help too. BTW, are you in the Discord?

Of course, my nickname is forceless!
Moreover, please check if there's anything else we should complete in this PR.

**Comment by edgarriba** [comment]
let just move the internal functions in a kernel.rs and good to go

**Comment by Force1ess** [comment]
> let just move the internal functions in a kernel.rs and good to go

Hi, I've implemented and placed them in the `kernel` crate.

**Comment by edgarriba** [comment]
ok, update then the tensor-ops to use the new kernels functions

**Review by edgarriba — COMMENTED** [review]
check that the test pass in you local machine

**Comment by Force1ess** [comment]
@edgarriba, I've  used the kernel functions and all tests passed in my machine

image

image

> check that the test pass in you local machine

File: crates/kornia-tensor-ops/src/error.rs [hunk 7]
```diff
@@ -11,4 +12,12 @@ pub enum TensorOpsError {
     /// Tensor error
     #[error("Error with the tensor: {0}")]
     TensorError(#[from] TensorError),
+
+    /// Tensor error
+    #[error("Error with the kernel: {0}")]
```
Comment (edgarriba) [inline crates/kornia-tensor-ops/src/error.rs:17]:
```suggestion
    #[transparent]
```
I recently learnt this trick, to forward directly the error

File: crates/kornia-tensor-ops/src/ops.rs [hunk 8]
```diff
@@ -151,4 +292,118 @@ mod tests {
...
+        let result = cosine_similarity(&h, &i)?;
+        assert!(result.abs() < 1e-6);
+        Ok(())
+    }
+
+    #[test]
+    fn test_cosine_similarity_3d_tensors() -> Result<(), TensorOpsError> {
+        let a = Tensor::<f32, 3, CpuAllocator>::from_shape_slice(
+            [2, 2, 2],
+            &[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0],
```
Comment (edgarriba) [inline crates/kornia-tensor-ops/src/ops.rs:367]: now seeing this test, I think we should limit also to Tensor1 this operator, as here i don't think it makes a lot of sense as it's applying to the whole tensor instead to a specific axis dimension ?

**Comment by edgarriba** [comment]
@Force1ess rebase from main as I fixed the ci issues

**Comment by Force1ess** [comment]
I've finished improvements as your reviews mentioned, and merged the latest `main` branch.
@edgarriba

**Review by edgarriba — APPROVED** [review]
